### ISSUE TITLE
Unexpected diagnostics and compiler crash when compiling FfiNative functions with dart2wasm

### ISSUE THREAD

**Issue by eyebrowsoffire** [dart-lang/sdk#50381]
I have switched over from wasm pragmas to @FfiNative annotations in dart2wasm and I'm running into a compiler crash during the WasmFfiNativeTransformer pass. It also shows up with some diagnostics that I wouldn't expect. Here is the full output:

```
lib/src/engine/skwasm/skwasm_impl/src/raw/raw_memory.dart:12:23: Error: Only JS interop members may be 'external'.
Try removing the 'external' keyword or adding a JS interop annotation.
external StackPointer stackAlloc(int length);
                      ^
lib/src/engine/skwasm/skwasm_impl/src/raw/raw_memory.dart:15:23: Error: Only JS interop members may be 'external'.
Try removing the 'external' keyword or adding a JS interop annotation.
external StackPointer stackSave();
                      ^
lib/src/engine/skwasm/skwasm_impl/src/raw/raw_memory.dart:18:15: Error: Only JS interop members may be 'external'.
Try removing the 'external' keyword or adding a JS interop annotation.
external void stackRestore(StackPointer pointer);
              ^
lib/src/engine/skwasm/skwasm_impl/src/raw/raw_memory.dart:12:23: Error: Only JS interop members may be 'external'.
Try removing the 'external' keyword or adding a JS interop annotation.
external StackPointer stackAlloc(int length);
                      ^
lib/src/engine/skwasm/skwasm_impl/src/raw/raw_memory.dart:15:23: Error: Only JS interop members may be 'external'.
Try removing the 'external' keyword or adding a JS interop annotation.
external StackPointer stackSave();
                      ^
lib/src/engine/skwasm/skwasm_impl/src/raw/raw_memory.dart:18:15: Error: Only JS interop members may be 'external'.
Try removing the 'external' keyword or adding a JS interop annotation.
external void stackRestore(StackPointer pointer);
              ^
lib/src/engine/skwasm/skwasm_impl/src/raw/raw_memory.dart:12:23: Error: Only JS interop members may be 'external'.
Try removing the 'external' keyword or adding a JS interop annotation.
external StackPointer stackAlloc(int length);
                      ^
lib/src/engine/skwasm/skwasm_impl/src/raw/raw_memory.dart:15:23: Error: Only JS interop members may be 'external'.
Try removing the 'external' keyword or adding a JS interop annotation.
external StackPointer stackSave();
                      ^
lib/src/engine/skwasm/skwasm_impl/src/raw/raw_memory.dart:18:15: Error: Only JS interop members may be 'external'.
Try removing the 'external' keyword or adding a JS interop annotation.
external void stackRestore(StackPointer pointer);
              ^
Unhandled exception:
Crash when compiling:
Null check operator used on a null value

#0      WasmFfiNativeTransformer._convertFfiTypeToWasmType (package:dart2wasm/ffi_native_transformer.dart:357)
#1      WasmFfiNativeTransformer.visitProcedure (package:dart2wasm/ffi_native_transformer.dart:174)
#2      Procedure.accept (package:kernel/ast.dart:3453)
#3      Transformer.transform (package:kernel/visitor.dart:1633)
#4      Transformer.transformList (package:kernel/visitor.dart:1668)
#5      Library.transformChildren (package:kernel/ast.dart:612)
#6      Transformer.defaultTreeNode (package:kernel/visitor.dart:1691)
#7      TreeVisitor.defaultExpression (package:kernel/visitor.dart:199)
#8      FfiTransformer.visitLibrary (package:vm/transformations/ffi/common.dart:516)
#9      _GrowableList.forEach (dart:core-patch/growable_array.dart:416)
#10     transformLibraries (package:dart2wasm/ffi_native_transformer.dart:49)
#11     WasmTarget.performModularTransformationsOnLibraries (package:dart2wasm/target.dart:174)
#12     KernelTarget.runBuildTransformations (package:front_end/src/fasta/kernel/kernel_target.dart:1621)
#13     KernelTarget.buildComponent.<anonymous closure> (package:front_end/src/fasta/kernel/kernel_target.dart:652)
<asynchronous suspension>
#14     withCrashReporting (package:front_end/src/fasta/crash.dart:136)
<asynchronous suspension>
#15     _buildInternal (package:front_end/src/kernel_generator_impl.dart:206)
<asynchronous suspension>
#16     withCrashReporting (package:front_end/src/fasta/crash.dart:136)
<asynchronous suspension>
#17     kernelForProgramInternal.<anonymous closure> (package:front_end/src/api_prototype/kernel_generator.dart:62)
<asynchronous suspension>
#18     kernelForProgramInternal (package:front_end/src/api_prototype/kernel_generator.dart:61)
<asynchronous suspension>
#19     kernelForProgram (package:front_end/src/api_prototype/kernel_generator.dart:50)
<asynchronous suspension>
#20     compileToModule (package:dart2wasm/compile.dart:63)
<asynchronous suspension>
#21     main (file:///opt/s/w/ir/x/w/sdk/pkg/dart2wasm/bin/dart2wasm.dart:120)
<asynchronous suspension>
Unhandled exception:
Crash when compiling:
Null check operator used on a null value

#0      WasmFfiNativeTransformer._convertFfiTypeToWasmType (package:dart2wasm/ffi_native_transformer.dart:357)
#1      WasmFfiNativeTransformer.visitProcedure (package:dart2wasm/ffi_native_transformer.dart:174)
#2      Procedure.accept (package:kernel/ast.dart:3453)
#3      Transformer.transform (package:kernel/visitor.dart:1633)
#4      Transformer.transformList (package:kernel/visitor.dart:1668)
#5      Library.transformChildren (package:kernel/ast.dart:612)
#6      Transformer.defaultTreeNode (package:kernel/visitor.dart:1691)
#7      TreeVisitor.defaultExpression (package:kernel/visitor.dart:199)
#8      FfiTransformer.visitLibrary (package:vm/transformations/ffi/common.dart:516)
#9      _GrowableList.forEach (dart:core-patch/growable_array.dart:416)
#10     transformLibraries (package:dart2wasm/ffi_native_transformer.dart:49)
#11     WasmTarget.performModularTransformationsOnLibraries (package:dart2wasm/target.dart:174)
#12     KernelTarget.runBuildTransformations (package:front_end/src/fasta/kernel/kernel_target.dart:1621)
#13     KernelTarget.buildComponent.<anonymous closure> (package:front_end/src/fasta/kernel/kernel_target.dart:652)
<asynchronous suspension>
#14     withCrashReporting (package:front_end/src/fasta/crash.dart:136)
<asynchronous suspension>
#15     _buildInternal (package:front_end/src/kernel_generator_impl.dart:206)
<asynchronous suspension>
#16     withCrashReporting (package:front_end/src/fasta/crash.dart:136)
<asynchronous suspension>
#17     kernelForProgramInternal.<anonymous closure> (package:front_end/src/api_prototype/kernel_generator.dart:62)
<asynchronous suspension>
#18     kernelForProgramInternal (package:front_end/src/api_prototype/kernel_generator.dart:61)
<asynchronous suspension>
#19     kernelForProgram (package:front_end/src/api_prototype/kernel_generator.dart:50)
<asynchronous suspension>
#20     compileToModule (package:dart2wasm/compile.dart:63)
<asynchronous suspension>
#21     main (file:///opt/s/w/ir/x/w/sdk/pkg/dart2wasm/bin/dart2wasm.dart:120)
<asynchronous suspension>
Unhandled exception:
Crash when compiling:
Null check operator used on a null value

#0      WasmFfiNativeTransformer._convertFfiTypeToWasmType (package:dart2wasm/ffi_native_transformer.dart:357)
#1      WasmFfiNativeTransformer.visitProcedure (package:dart2wasm/ffi_native_transformer.dart:174)
#2      Procedure.accept (package:kernel/ast.dart:3453)
#3      Transformer.transform (package:kernel/visitor.dart:1633)
#4      Transformer.transformList (package:kernel/visitor.dart:1668)
#5      Library.transformChildren (package:kernel/ast.dart:612)
#6      Transformer.defaultTreeNode (package:kernel/visitor.dart:1691)
#7      TreeVisitor.defaultExpression (package:kernel/visitor.dart:199)
#8      FfiTransformer.visitLibrary (package:vm/transformations/ffi/common.dart:516)
#9      _GrowableList.forEach (dart:core-patch/growable_array.dart:416)
#10     transformLibraries (package:dart2wasm/ffi_native_transformer.dart:49)
#11     WasmTarget.performModularTransformationsOnLibraries (package:dart2wasm/target.dart:174)
#12     KernelTarget.runBuildTransformations (package:front_end/src/fasta/kernel/kernel_target.dart:1621)
#13     KernelTarget.buildComponent.<anonymous closure> (package:front_end/src/fasta/kernel/kernel_target.dart:652)
<asynchronous suspension>
#14     withCrashReporting (package:front_end/src/fasta/crash.dart:136)
<asynchronous suspension>
#15     _buildInternal (package:front_end/src/kernel_generator_impl.dart:206)
<asynchronous suspension>
#16     withCrashReporting (package:front_end/src/fasta/crash.dart:136)
<asynchronous suspension>
#17     kernelForProgramInternal.<anonymous closure> (package:front_end/src/api_prototype/kernel_generator.dart:62)
<asynchronous suspension>
#18     kernelForProgramInternal (package:front_end/src/api_prototype/kernel_generator.dart:61)
<asynchronous suspension>
#19     kernelForProgram (package:front_end/src/api_prototype/kernel_generator.dart:50)
<asynchronous suspension>
#20     compileToModule (package:dart2wasm/compile.dart:63)
<asynchronous suspension>
#21     main (file:///opt/s/w/ir/x/w/sdk/pkg/dart2wasm/bin/dart2wasm.dart:120)
<asynchronous suspension>
description: Sub-process failed.executable: /Users/jacksongardner/Source/flutter/engine/src/flutter/prebuilts/macos-arm64/dart-sdk/bin/dartaotruntime arguments: [/Users/jacksongardner/Source/flutter/engine/src/flutter/prebuilts/macos-arm64/dart-sdk/bin/snapshots/dart2wasm_product.snapshot, --dart-sdk=/Users/jacksongardner/Source/flutter/engine/src/flutter/prebuilts/macos-arm64/dart-sdk, -DFLUTTER_WEB_AUTO_DETECT=false, -DFLUTTER_WEB_USE_SKIA=false, -DFLUTTER_WEB_USE_SKWASM=true, test/ui/color_test.dart, /Users/jacksongardner/Source/flutter/engine/src/flutter/lib/web_ui/build/test/ui/color_test.dart.browser_test.dart.wasm] exit code: 255
```
I wouldn't expect the "Only JS interop members may be external" diagnostic, and weirdly it's only on one of the files that uses the FfiNative annotations.

For a minimum reproduction, you can check out my branch: https://github.com/eyebrowsoffire/engine/tree/skwasm_renderer

The branch is half baked but you should be able to invoke the compiler on the test in question from my branch. The command should look something like this:

```
dart dart2wasm.dart --dart-sdk=<path to dart sdk> -DFLUTTER_WEB_AUTO_DETECT=false -DFLUTTER_WEB_USE_SKIA=false -DFLUTTER_WEB_USE_SKWASM=true <path to my flutter branch>/lib/web_ui/test/ui/color_test.dart
```

**Comment by joshualitt** [comment]
@osa1 Apparently the crash and the validation error are unrelated. I have a fix for the validation error which should land soon.

**Comment by eyebrowsoffire** [comment]
In my `@FfiNative` annotations, I am using ABI specific numeric native types, such as `Int`, `Size`, etc. These can be found in `dart-sdk/lib/ffi/c_type.dart`. The analyzer seems to accept these as valid native types for the purposes of the annotation's arguments and return values, but `WasmFfiNativeTransformer` crashes when processing them. The `WasmFfiNativeTransformer` calls `getType` on each parameter and return value in the annotation, and `getType` comes from one of its parent classes, `FfiTransformer`. `FfiTransformer` keeps a mapping called `classNativeTypes` that map the classes themselves back to an enum that represents the possible types.

However, this mapping does *not* contain the ABI specific numeric native types, it only has numeric types of a known width. I think this is intentional, which says to me that probably before calling `getType` we need to convert the ABI specific type to its fixed width type, and *then* call `getType` on it. Each of these ABI specific numeric native types themselves has an `@AbiSpecificMapping` annotation that tells what fixed-width type it corresponds to on each platform. So I just have to find a utility that uses that mapping and use it into the transformer before we call `getType`.

**Comment by eyebrowsoffire** [comment]
Fix for the crash is out for review too: https://dart-review.googlesource.com/c/sdk/+/268105/1

**Comment by osa1** [comment]
Fixed with 81a63dafe533f586d77ecf4688a88695423e2286.
